devito/types/dense.py

```diff
@@ -959,7 +959,8 @@ def __init_finalize__(self, *args, **kwargs):
             raise TypeError("`space_order` must be int or 3-tuple of ints")
 
         # Dynamically add derivative short-cuts
-        self._fd = generate_fd_shortcuts(self)
+        self._fd = generate_fd_shortcuts(self.dimensions, self.space_order,
+                                         to=getattr(self, 'time_order', 0))
```

Comment:
this is a bit horrible. The OO way to do this is to use methods and override them in the subclass

```
class Function(...):

     def _generate_fd_shortcuts(self):
            return generate_fd_shortcuts(self.dimensions, self.space_order)


class TimeFunction(...):

          def _generate_fd_shortcuts(self):
                return generate_fd_shortcuts(self.dimensions, self.space_order, self.time_order)
```

Reply:
in fact, we should stick to our hooks convention, and call it something like `__fd_setup__`